tools/triton-sme-opt/triton-sme-opt.cpp

```diff
@@ -0,0 +1,178 @@
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/InitAllDialects.h"
+#include "mlir/InitAllPasses.h"
+#include "mlir/Interfaces/VectorInterfaces.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Tools/mlir-opt/MlirOptMain.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include <iostream>
+#include "mlir/IR/Operation.h"
+#include <llvm/Support/raw_ostream.h>
+
+
+using namespace mlir;
+
+
+namespace {
+struct OuterProductVectorizationPass
+    : public PassWrapper<OuterProductVectorizationPass,
+                         OperationPass<func::FuncOp>> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<vector::VectorDialect, func::FuncDialect>();
+  }
+
+  void runOnOperation() override {
+    func::FuncOp funcOp = getOperation();
+    MLIRContext *context = funcOp.getContext();
+    RewritePatternSet patterns(context);
+    ConversionTarget target(*context);
+
+      // Apply patterns for lowering masked transfers
+    transform::ApplyLowerMaskedTransfersPatternsOp lowerMaskedTransfersPatterns;
+    lowerMaskedTransfersPatterns.populatePatterns(patterns);
+
+    // Apply patterns for transfer permutation
+    transform::ApplyTransferPermutationPatternsOp transferPermutationPatterns;
+    transferPermutationPatterns.populatePatterns(patterns);
+
+    // Apply patterns for reduction to contract
+    transform::ApplyVectorReductionToContractPatternsOp reductionToContractPatterns;
+    reductionToContractPatterns.populatePatterns(patterns);
+
```

Comment:
You are missing patterns from `vector.contract` to `vector.outerproduct`:
* https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Vector/TransformOps/VectorTransformOps.td#L106

Reply:
@nhat-nguyen turns out that the whole kernel need to be bufferized before we can run this pass. Can we use structured-to-memref todo this? would make things a lot easier

Reply:
yeah you can use a combination of of `--triton-to-structured --canonicalize --triton-arith-to-linalg --structured-to-memref`, this is the equivalent of `--triton-to-linalg`. Although this only converts the loads and stores to memref, you would need to also run the bufferization pass to convert the remaining ops to use memref too.

Reply:
yea i was a little worried about this, looks like all the tensor ops will have to be lowered too

Reply:
@danikhan632 You can try creating a new pass that wraps the bufferization pass and use the following. This is what we have been using.

```cpp
  void runOnOperation() final {
    auto options = mlir::bufferization::OneShotBufferizationOptions();
    options.allowReturnAllocsFromLoops = true;
    options.allowUnknownOps = true;
    options.bufferAlignment = 16;
    options.bufferizeFunctionBoundaries = true;
    options.unknownTypeConverterFn =
        [](mlir::Value value, mlir::Attribute memorySpace,
           const mlir::bufferization::BufferizationOptions &options) {
          return mlir::bufferization::getMemRefTypeWithStaticIdentityLayout(
              value.getType().cast<mlir::TensorType>(), memorySpace);
        };
    options.setFunctionBoundaryTypeConversion(
        mlir::bufferization::LayoutMapOption::IdentityLayoutMap);
    options.getMemorySpaceFn = [](mlir::TensorType t) {
      if (auto rt = t.dyn_cast<mlir::RankedTensorType>())
        return rt.getEncoding();
      return mlir::Attribute();
    };

    mlir::bufferization::BufferizationStatistics stats;
    auto moduleOp = getOperation();
    if (failed(mlir::bufferization::runOneShotModuleBufferize(moduleOp, options,
                                                              &stats))) {
      signalPassFailure();
    }

    // copied from OneShotBufferizePass to keep output IR consistent
    OpPassManager cleanupPipeline("builtin.module");
    cleanupPipeline.addPass(createCanonicalizerPass());
    cleanupPipeline.addPass(createCSEPass());
    cleanupPipeline.addPass(createLoopInvariantCodeMotionPass());
    (void)runPipeline(cleanupPipeline, moduleOp);
  }
```

Reply:
@banach-space We have a mix of tensor + memref semantics, but the memrefs are only at the load and store boundaries. Assume you have the following:

```python
a = tl.load(in_ptr + tl.arange(0, 8))
b = a * 2 + 1
c = b / 2
tl.store(out_ptr + tl.arange(0, 8))
```

We first assume that the data each triton program instance uses will first be copied to its local memory, so `tl.load` is lowered to:

```mlir
%memref = memref.reinterpret_cast %in_ptr with the correct offset and strides
%local_buffer = memref.alloc
memref.copy %memref %local_buffer
%tensor = bufferization.to_tensor %local_buffer
```

All data transformation ops will then operate on `%tensor`, this means linalg tiling on tensors will work as intended.

And finally, before storing the data back, we lower `tl.store` to:

```
%memref_out = memref.reinterpret_cast %out_ptr with the correct offset and strides
bufferization.materialize_in_destination %some_tensor %memref_out
```

With this semantic, we don't have any requirements on when kernels produced by `--triton-to-linalg` should be bufferized. And like you mentioned, we do in fact delay bufferization as late as possible and perform many transformations before that. Hope this helps!

Reply:
from what I'm hearing, bufferizing might be necessary before step #2 

```
         (1)    bufferize here     (2)                      (3)
 linalg.matmul -------------> vector.contract ----> vector.outerproduct -----> arm_sme.fmopa
```



in that case optimizations may be difficult todo optimizations. Might still write a bufferization pass to make the compiler happy but def not optimal

@banach-space if I wanted to make future optizations could I modify the lowering to also accept tensors and write some lowering logic? Been thinking it over and bufferizing too early could be bad for performance and given that SME is a matmul engine I'd imagine being able to accept tensors might be useful

Reply:
> from what I'm hearing, bufferizing might be necessary before step 2

I don't think so, that's still quite high level and there wouldn't be any SME ops yet :) 

> Been thinking it over and bufferizing too early could be bad for performance and given that SME is a matmul engine I'd imagine being able to accept tensors might be useful

In my view Tensor is too high level/abstract type for the ArmSME dialect. In particular, the hardware won't know what a tensor is. I suggest the following:
```
     (1)                    (2)                 (3)              bufferize here
 linalg.matmul -----> vector.contract ----> vector.outerproduct -----------------> arm_sme.fmopa
```

This would be consistent with what we currently do in MLIR:
* https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/multi-tile-matmul.mlir

As you can see, bufferization happens after lowering `vector.contract` to `vector.outerproduct`.

Btw, I deliberately made the distinction into `vector.contract` and `vector.outerproduct` as that's representative of how the Vector Dialect splits Ops into "high level" (e.g. `vector.contract`) and "low level" (e.g. `vector.outerproduct`) vector ops. Here's an overview of the vectoriser n the context of SME:
* https://www.youtube.com/watch?v=jHk56V6cHN8&ab_channel=LLVM

Thanks again for working on this - hope this helps :)

Reply:
The video makes sense but a just want to clarify how the passes would be handled
This is what should happen

```cpp
  PassPipelineRegistration<> smeConversionPipeline(
      "sme-conversion",
      "Converts linalg.matmul to a more optimized form using SME",
      [](OpPassManager &pm) {
        pm.addPass(createMatmulTileConversionPass(true));   //tiling and vectorizing  linalg.matmul
        pm.addPass(createOuterProductVectorizationPass());  // lowering vector.contract to vector.outerproduct
        pm.addPass(createLinalgBufferizePass()); //bufferization happens here
      });

```
let me know if this is accurate

and ofc happy to work on this :)

Reply:
This makes sense to me, but I haven't checked the names of the actual passes. Also, you may need to configure some of these passes to do what you want.

I suggest implementing this step-by-step. At every step you can compare against the output that the Transform Dialect sequence [generates](https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/multi-tile-matmul.mlir) - if there's divergence then that might mean trouble :) (but not necessarily)

Note that in your list of passes there's nothing SVE/SME specific, so you won't be lowering to SME just yet. That's fine - baby steps ;-)

Reply:
just updated, it looks mostly bufferized